### PR TITLE
fix: review workflow persist data

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

Review workflows adds two additional attributes to content types, strapi_assignee and strapi_stage . Those are the values displayed in the content manager edit view.

As RW is an EE feature, it was decided for those attributes to not exist on the Comunity Edition. But at the same time if a user filled those values in EE and they downgrade, we want to keep the data there.

In v4 we were using a persist_table logic, to persist those join tables even if the attribute didn't exist. At some point we discussed doing otherwise, but eventually this was forgotten.

This PR reintroduces the persisted table logic we had in in v4.

### CODE REVIEW

File: packages/core/review-workflows/server/src/register.ts [hunk 1]
```diff
@@ -56,33 +54,62 @@ function extendReviewWorkflowContentTypes({ strapi }: { strapi: Core.Strapi }) {
...
+/**
+ * Persist the stage & assignee attributes so they are not removed when downgrading to CE.
+ *
+ * TODO: V6 - Instead of persisting the join tables, always create the stage & assignee attributes, even in CE mode
+ *            It was decided in V4 & V5 to not expose them in CE (as they pollute the CTs) but it's not worth given the complexity this needs
+ */
+function persistRWOnDowngrade({ strapi }: { strapi: Core.Strapi }) {
+  const { removePersistedTablesWithSuffix, persistTables } = getAdminService('persist-tables');
+
+  return async ({ contentTypes }: any) => {
```
Comment (markkaylor) [inline packages/core/review-workflows/server/src/register.ts:66]: We don't have any types for contentTypes?

File: packages/core/review-workflows/server/src/register.ts [hunk 2]
```diff
@@ -56,33 +54,62 @@ function extendReviewWorkflowContentTypes({ strapi }: { strapi: Core.Strapi }) {
...
+    const getStageTableToPersist = (contentTypeUID: any) => {
```
Comment (markkaylor) [inline packages/core/review-workflows/server/src/register.ts:67]: Don't we know that the uid is a string?

File: packages/core/review-workflows/server/src/register.ts [hunk 3]
```diff
@@ -56,33 +54,62 @@ function extendReviewWorkflowContentTypes({ strapi }: { strapi: Core.Strapi }) {
...
+      const joinTableName = attributes[ENTITY_ASSIGNEE_ATTRIBUTE].joinTable.name;
+      return {
+        name: joinTableName,
+        dependsOn: [{ name: tableName }],
+      };
+    };
+
+    const stageJoinTablesToPersist = pipe([
+      getVisibleContentTypesUID,
+      filter((uid: any) => hasStageAttribute(contentTypes[uid])),
```
Comment (markkaylor) [inline packages/core/review-workflows/server/src/register.ts:89]: string? 🤔

File: packages/core/review-workflows/server/src/register.ts [hunk 4]
```diff
@@ -56,33 +54,62 @@ function extendReviewWorkflowContentTypes({ strapi }: { strapi: Core.Strapi }) {
...
+
+    // Remove previously created join tables and persist the new ones
+    await removePersistedTablesWithSuffix('_strapi_stage_lnk');
+    await persistTables(stageJoinTablesToPersist);
+
+    const assigneeJoinTablesToPersist = pipe([
+      getVisibleContentTypesUID,
+      filter((uid: any) => hasStageAttribute(contentTypes[uid])),
+      map(getAssigneeTableToPersist),
+    ])(contentTypes);
```
[the remark this answers is not in the document]
Comment (markkaylor) [reply packages/core/review-workflows/server/src/register.ts:101]: I may not understand exactly what you have in mind @remidej , but for what it's worth, I am personally not too bothered by the repetition here. It's all in one place, readable, and easy to follow. I feel like an abstraction might take away from that just to be more concise.

File: packages/core/review-workflows/server/src/register.ts [hunk 5]
```diff
@@ -56,33 +54,62 @@ function extendReviewWorkflowContentTypes({ strapi }: { strapi: Core.Strapi }) {
...
+
+    const stageJoinTablesToPersist = pipe([
+      getVisibleContentTypesUID,
+      filter((uid: any) => hasStageAttribute(contentTypes[uid])),
+      map(getStageTableToPersist),
+    ])(contentTypes);
+
+    // Remove previously created join tables and persist the new ones
+    await removePersistedTablesWithSuffix('_strapi_stage_lnk');
+    await persistTables(stageJoinTablesToPersist);
```
[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/review-workflows/server/src/register.ts:95]: indeed, will add a comment to make it clearer.
There might be content types that have been removed or modified, hence the persisted table list has to be recreated again.

Its inefficient, and hacky. But its the only way we found to make this work at the time

File: packages/core/review-workflows/server/src/register.ts [hunk 6]
```diff
@@ -56,33 +54,62 @@ function extendReviewWorkflowContentTypes({ strapi }: { strapi: Core.Strapi }) {
...
-//     };
-
-//     const joinTablesToPersist = pipe([
-//       getVisibleContentTypesUID,
-//       filter((uid: any) => hasStageAttribute(contentTypes[uid])),
-//       map(getStageTableToPersist),
-//     ])(contentTypes);
-
-// await removePersistedTablesWithSuffix('_strapi_stage_links');
-// await persistTables(joinTablesToPersist);
```
[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/review-workflows/server/src/register.ts:79]: nope, looks like we were not doing that in v4 either

File: packages/core/review-workflows/server/src/register.ts (hunk 1 above)
Comment (Marc-Roig) [reply packages/core/review-workflows/server/src/register.ts:66]: we do! will add it :)

File: packages/core/review-workflows/server/src/register.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/review-workflows/server/src/register.ts:101]: :cheffkiss:

File: packages/core/admin/ee/server/src/services/persist-tables.ts [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import type { Core } from '@strapi/types';
 import { differenceWith, isEqual } from 'lodash/fp';
 
-interface PersistedTable {
+export interface PersistedTable {
```
Comment (markkaylor) [inline packages/core/admin/ee/server/src/services/persist-tables.ts:4]: I just don't see where you are using this outside this file 🤔

**Review by markkaylor — APPROVED** [review]
Nice, looks good to me.

File: packages/core/admin/ee/server/src/services/persist-tables.ts (hunk 7 above)
Comment (Marc-Roig) [reply packages/core/admin/ee/server/src/services/persist-tables.ts:4]: typescript was complaining about it, if another file is importing a function that references this type, it also needs to be exported to be visible (or thats what I was able to understand)

**Pull request merged**